proposals/p2173.md

```diff
@@ -0,0 +1,616 @@
+on the language to avoid the hard cases. Some languages merely impose some kind
+of depth limit on their search for a proof of equality.
+
+Note that if we can perform type canonicalization, then type equality is
+necessarily transitive: if all equal _pairs_ of types canonicalize to the same
+representation, then _all_ equal types canonicalize to the same representation
+transitively.
+
+### Coherence
+
+In order to avoid surprises in the language behavior, we want Carbon to have the
+property that learning more about the type of a value should not cause the
+behavior of an expression to change from one valid behavior to another valid
+behavior.
+
+It's important to distinguish here between learning more about a type, such as
+learning that it implements a particular interface – for example, from an added
+import – and changing the type, such as by writing a different expression after
+a `:` or `:!`. Changing the type of a value is expected to be able to change the
+behavior of an expression using that value.
+
+## Background
+
+There has been a lot of work in this space. Some previous relevant proposals:
+
+-   [#731: generics details 2](https://github.com/carbon-language/carbon-lang/pull/731)
+    introduced associated types, with the intent that a mechanism to constrain
+    their value would be provided in a later proposal.
+-   [#818: generics details 3](https://github.com/carbon-language/carbon-lang/pull/818)
+    introduced `where` constraints for generics, with an open question for
+    whether and how we might restrict constraints in order to give a decidable
+    type equality rule.
+    -   #818 represents the status quo of approved proposals. It has both `=`
+        and `==` where constraints, which are similar to, but somewhat different
+        from, those in this proposal.
+    -   `=` constraints require a concrete type on their right-hand side. The
+        left-hand side is not restricted. The type-of-type of the left-hand side
+        is changed to match the right-hand side.
+    -   `==` constraints do not affect the type-of-type of either operand.
+    -   As an ansewr to the question of how to produce a decidable type equality
+        rule, both kinds of constraints are applied automatically, but only at a
+        depth of one constraint. Neither form is transitive but both can be
+        transitively extended using `observe` declarations.
+-   [#2070: always `==` not `=` in `where` clauses](https://github.com/carbon-language/carbon-lang/pull/2070)
+    describes an attempt to unify the syntax and semantics of `=` and `==`
+    constraints, under which we would always merge the type-of-types of the
+    operands. That proposal has not yet been accepted, and this proposal intends
+    to supersede it.
+
+## Proposal
+
+Replace the existing `where A == B` constraint with two different kinds of
+constraint:
+
+-   `where .A = T` specifies a _rewrite_ constraint. If the associated constant
```

Comment:
I agree that this can lead to some undesirable syntactic forms (`let T:! Foo where .A = i32 = i32`). Added some "Alternatives considered" text, but I don't have a concrete preferable alternative so I'm not planning on addressing this problem in this proposal. This is also not a new problem with this proposal, so I think we shouldn't couple this proposal to solving this problem.